### PR TITLE
Clean up bmf.cpp

### PR DESCRIPTION
**Pull request by miller-alex**
When I started working on the fixes that were merged recently, I did some additional formatting changes and refactoring for the first file I was working on. (I didn't do that for other files because there's a lot more important stuff to fix.) Since these commits introduce no functional change and don't fix anything I didn't include them in the other pull requests.

Some of the changes may be a matter of taste, but I think most of them improve readability of the code. Feel free to merge if you like the changes (or modify what you don't like or even close without merging). What do you think?

### CODE REVIEW

File: src/bmf.cpp [hunk 1]
```diff
@@ -83,23 +87,25 @@ bool CxadbmfPlayer::xadplayer_load()
...
+  if (xad.fmt != BMF)
     return false;
 
 #ifdef DEBUG
   AdPlug_LogWrite("\nbmf_load():\n\n");
 #endif
   if (tune_size < 6)
     return false;
-  if (!strncmp((char *)&tune[0],"BMF1.2",6))
+  if (!strncmp((char *)tune, "BMF1.2", 6))
```
Comment (Malvineous) [inline src/bmf.cpp:98]: This one can sometimes be preferable the way it was, as it makes it a bit clearer you're talking about offset 0 in the file.

File: src/bmf.cpp [hunk 2]
```diff
@@ -83,23 +87,25 @@ bool CxadbmfPlayer::xadplayer_load()
...
   {
     bmf.version = BMF1_2;
     bmf.timer = 70.0f;
+    ptr = 6;
   }
-  else if (!strncmp((char *)&tune[0],"BMF1.1",6))
+  else if (!strncmp((char *)tune, "BMF1.1", 6))
   {
```
Comment (Malvineous) [inline src/bmf.cpp:105]: Same here.  It's probably a matter of taste, but as we're often looking at files in a hex editor when implementing this stuff I do like the added clarity you can get from `[0]`.

File: src/bmf.cpp [hunk 3]
```diff
@@ -141,48 +145,46 @@ bool CxadbmfPlayer::xadplayer_load()
...
       if (iflags & (1 << (31-i)))
-	  {
+      {
         if (tune_size - ptr < 24)
           return false;
         strncpy(bmf.instruments[i].name, (char *)tune + ptr, 10);
         // 11th char ignored since name must be 0-terminated
         bmf.instruments[i].name[10] = 0;
-        memcpy(bmf.instruments[i].data, &tune[ptr+11], 13);
+        memcpy(bmf.instruments[i].data, tune + ptr + 11, 13);
```
Comment (Malvineous) [inline src/bmf.cpp:177]: Same argument here.  Given that `[]` is used a lot elsewhere, I think keeping `[]` in use makes the code a bit more uniform.  Otherwise you could make the argument that on line 161 above, `(tune[ptr] << 24) | (tune[ptr + 1] << 16)` should be replaced with the horrible looking `(*(tune + ptr) << 24) | (*(tune + ptr + 1) << 16))` to match the new change here :)

File: src/bmf.cpp [hunk 4]
```diff
@@ -141,48 +145,46 @@ bool CxadbmfPlayer::xadplayer_load()
...
         ptr += 24;
-	  }
+      }
+      else if (bmf.version == BMF1_1)
+      {
+        memset(bmf.instruments[i].name, 0, 11);
+        memcpy(bmf.instruments[i].data, bmf_default_instrument, 13);
```
Comment (Malvineous) [inline src/bmf.cpp:183]: Is there any way you can replace the hard-coded constants with something more future-proof like `sizeof(bmf_default_instrument)`?

File: src/bmf.cpp [hunk 5]
```diff
@@ -141,48 +145,46 @@ bool CxadbmfPlayer::xadplayer_load()
...
-        bmf.instruments[i].name[0] = 0;
-		
-        if (bmf.version == BMF1_1)
-		  for(int j=0;j<13;j++)
-			bmf.instruments[i].data[j] = bmf_default_instrument[j];
-        else
-		  for(int j=0;j<13;j++)
-			bmf.instruments[i].data[j] = 0;
-	  }
+        memset(bmf.instruments + i, 0, sizeof(bmf.instruments[i]));
```
Comment (Malvineous) [inline src/bmf.cpp:186]: Yes like `sizeof()` here, this is good.

File: src/bmf.cpp [hunk 6]
```diff
@@ -204,16 +206,17 @@ bool CxadbmfPlayer::xadplayer_load()
...
+
   // load streams
   if (bmf.version > BMF0_9B)
   {
     if (tune_size - ptr < 4)
       return false;
-    unsigned long sflags = (tune[ptr] << 24) | (tune[ptr+1] << 16) | (tune[ptr+2] << 8) | tune[ptr+3];
-    ptr+=4;
+    unsigned long sflags = (tune[ptr] << 24) | (tune[ptr+1] << 16)
+                         | (tune[ptr+2] << 8) | tune[ptr+3];
```
Comment (Malvineous) [inline src/bmf.cpp:216]: Does this look neater if rather than shortening it for the sake of line length, you just put each `tune[]` thing on its own line?  Might make it easier to see the pattern?

File: src/bmf.cpp [hunk 7]
```diff
@@ -266,162 +261,159 @@ void CxadbmfPlayer::xadplayer_rewind(int subsong)
...
-        memcpy(&event, &bmf.streams[i][bmf.channel[i].stream_position], sizeof(bmf_event));
+again:
+    const bmf_event &event = bmf.streams[i][pos];
 #ifdef DEBUG
-   AdPlug_LogWrite("%02X %02X %02X %02X %02X %02X\n",
-		   event.note,event.delay,event.volume,event.instrument,
-		   event.cmd,event.cmd_data);
+    AdPlug_LogWrite("%02X %02X %02X %02X %02X %02X\n",
+                    event.note, event.delay, event.volume,
+                    event.instrument, event.cmd, event.cmd_data);
```
Comment (Malvineous) [inline src/bmf.cpp:305]: Personally I'm not a fan of excessive indenting without a very good reason, I'd be happy with normal indenting here.

File: src/bmf.cpp [hunk 8]
```diff
@@ -467,193 +459,153 @@ long int CxadbmfPlayer::__bmf_convert_stream(const unsigned char *stream,
...
+  // until the end marker is found and stored.
+  for (int pos = 0; pos <= last_pos; pos < last_pos && pos++)
   {
-    if (pos >= sizeof(bmf.streams[0]) / sizeof(bmf.streams[0][0]))
-    {
-      // Stream too long to store.
-      pos--; // Truncate but keep reading until end marker
-      // or: return -1;
-    }
+    bmf_event &event = bmf.streams[channel][pos];
```
Comment (Malvineous) [inline src/bmf.cpp:475]: Always nice to shorten lengthy variable names, I like this.

File: src/bmf.cpp [hunk 9]
```diff
@@ -467,193 +459,153 @@ long int CxadbmfPlayer::__bmf_convert_stream(const unsigned char *stream,
...
       {
-        // byte0: 0aaaaaaa = NOTE
-        bmf.streams[channel][pos].note = *stream & 0x7F;
+        // 2nd byte is 1xDDDDDD: delay byte present
+        event.delay = *stream & 0x3F;
 
-		stream++;
-      } // if (*stream & 0x80)
-    } // if (*stream == 0xFE)
+        if (!(*(stream++) & 0x40))
```
Comment (Malvineous) [inline src/bmf.cpp:526]: Please confirm but I believe you can get away without the brackets here and in the other places: `*stream++` (as the `++` doesn't apply until the whole thing has finished being evaluated).

File: src/bmf.cpp (hunk 1 above)
Comment (miller-alex) [reply src/bmf.cpp:98]: It's a matter of taste, as I wrote above. While I don't like it very much, it's OK for me to restore the old `&[]` pattern if you prefer that. Just tell me whether you want me to replace the old commits or add a new one.
(Unrelated side note: I'd also prefer memcmp over strncmp here, but it's really minor and some people might not like it either.)

File: src/bmf.cpp (hunk 3 above)
Comment (miller-alex) [reply src/bmf.cpp:177]: See above. But to be clear, I don't want to ban `[]`, I just don't like `&[]`. No sane person would write `*(tune + ptr)`, and I've even changed some occurrences of `*(stream+1)` to `stream[1]` further down (although I can see why one might prefer the other form in that particular contex; and now `stream[1]` is used next to `*stream`, which is inconsistent, but `*stream` is a better fit than `stream[0]` there; whatever you do, you have to live with one ugly wart). Some people want conistency, others use different syntax depending on context to emphasize certain aspects, it's all complicated. I don't have a very strong opinion on this matter, just tell me what you want.

File: src/bmf.cpp (hunk 4 above)
Comment (miller-alex) [reply src/bmf.cpp:183]: Replacing literal constants with `sizeof` (or an `ARRAY_SIZE()` macro, `enum`, `#define`d constant) is always a good idea. But literal numbers are used everywhere, so I didn't bother. I can change it here if you want, but a general literal constant cleanup pass can (and should) be done sometime later. I'd rather fix some real bugs now.

File: src/bmf.cpp (hunk 6 above)
Comment (miller-alex) [reply src/bmf.cpp:216]: Actually this should not be open coded at all. This is a pattern that's used all over the place and many players have their own macros/functions dealing with it in addition to libbinio providing a solution for file data. All that should be consolidated into a common set of helpers used everywhere. But that's out of scope for this PR.

File: src/bmf.cpp (hunk 7 above)
Comment (miller-alex) [reply src/bmf.cpp:305]: I think most LogWrites are formatted like this, but I really don't care.

File: src/bmf.cpp (hunk 9 above)
Comment (miller-alex) [reply src/bmf.cpp:526]: Yes, the parentheses can be omitted since postfix operators have higher precedence than prefix unary operators, but for some people that's hard to memorize. I like the added clarity of the patentheses, but don't like the cluttered display, so it's a mixed bag, again.

**Comment by miller-alex** [comment]
I don't want to invest too much time into style changes. These are changes that have been sitting in my local repo for a while an I thought I should push them upstream. Originally I mostly wanted to simplify the control flow that was hard to follow, and in preparation I also cleaned up a few other things that I wouldn't have bothered with otherwise.

There are a lot of things that could be cleaned up or changed to a more pleasant style, but I suppose my time is better spent fixing real bugs. I might simplify or refactor some code, but style changes have to wait.

Let me know if/what you want me edit/add, or just change it yourself and merge it, no big deal.

**Comment by miller-alex** [comment]
@Malvineous, I've added patches with most of your suggestions. I think it's ready to merge unless you have more requests.

File: src/bmf.cpp (hunk 1 above)
Comment (Malvineous) [reply src/bmf.cpp]: I don't mind `strncmp` vs `memcmp` but since we're dealing with binary data here rather than text, `memcmp` seems a bit more logical.  It also gets rid of the typecast so happy with that change.

File: src/bmf.cpp (hunk 3 above)
Comment (Malvineous) [reply src/bmf.cpp]: Generally my opinion is that `&[]` is clearer to me when talking about offsets in files, as it's obvious you're looking at an offset because of the array syntax.  Something like `tune + 11` requires more mental effort for me, as I have to remember whether `tune` is being used as a pointer or an integer.

File: src/bmf.cpp (hunk 4 above)
Comment (Malvineous) [reply src/bmf.cpp]: Fair enough, I just figured if you were changing those lines anyway it'd be a good opportunity to fix it while you were there.

File: src/bmf.cpp (hunk 6 above)
Comment (Malvineous) [reply src/bmf.cpp]: Fair enough!

File: src/bmf.cpp (hunk 9 above)
Comment (Malvineous) [reply src/bmf.cpp]: You make a fair point, but by the same token memorising the operator precedence is probably going to serve someone well in the long run :)

**Comment by Malvineous** [comment]
All good, thanks for helping to improve the code!
